sklearn/utils/tests/test_validation.py

```diff
@@ -2358,3 +2358,24 @@ def test_force_all_finite_rename_warning():
 
     with pytest.warns(FutureWarning, match=msg):
         as_float_array(X, force_all_finite=True)
+
+
+@pytest.mark.parametrize(
+    ["X", "estimator", "expected_error_message"],
+    [
+        (
+            np.array([[[1, 2], [3, 4]], [[1, 2], [3, 4]]]),
+            RandomForestRegressor(),
+            "Found array with dim 3, while dim <= 2 is required by "
+            "RandomForestRegressor.",
+        ),
+        (
+            np.array([[[1, 2], [3, 4]], [[1, 2], [3, 4]]]),
+            None,
+            "Found array with dim 3, while dim <= 2 is required.",
+        ),
+    ],
+)
+def test_check_array_allow_nd_errors(X, estimator, expected_error_message):
+    with pytest.raises(ValueError, match=expected_error_message):
+        check_array(X, estimator=estimator, allow_nd=False)
```

Comment:
Since `allow_nd=False` by default, I would not specify it explictly here:
```suggestion
        check_array(X, estimator=estimator)
```

In other words, looking too quickly at your PR I thought that the issue only happened in the "special" case `allow_nd=False` and not the default case. Turn out this not so great error message was happening in the default case ...